### PR TITLE
feat: add multi-criteria support beyond 4K

### PR DESCRIPTION
**Pull request by dabigc**
## Summary

- **Genericize filtarr** to support multiple release criteria beyond just 4K
- **Add new criteria:** SPECIAL_EDITION (detects Special Edition, Collector's Edition, Anniversary Edition, Ultimate Edition, Definitive Edition)
- **Movie-only enforcement:** Directors-cut, extended, remaster, imax, special-edition only work for movies (raises error for TV series)
- **Pattern-based tagging:** Tags now follow `{criteria}-available` pattern (e.g., `imax-available`, `directors-cut-unavailable`)
- **CLI `--criteria` flag:** New option for `check movie`, `check series`, and `check batch` commands
- **Documentation updates:** Reposition as "release criteria filtering tool" with 4K as primary but not only use case

### Code Quality Improvements

- **Granular exception handling:** Replace broad `Exception` catches with specific `httpx.HTTPStatusError`, `httpx.ConnectError`, `httpx.TimeoutException`, and `pydantic.ValidationError`
- **Config refactoring:** Extract 12 helper functions for improved readability and maintainability
- **CLI refactoring:** Extract batch processing into focused helper functions with `BatchContext` dataclass

### Test Coverage Improvements

Added **115 new tests** to address Codecov coverage gaps, organized in 3 phases:

**Phase 1 - Quick Wins (71 tests):**
- `TestParseBatchFile`: Batch file parsing edge cases (18 tests)
- `TestCheckMovieCriteriaValidation`: Movie criteria validation (3 tests)
- `TestCheckSeriesCriteriaValidation`: Series criteria + movie-only errors (11 tests)
- `TestValidateBatchInputs`: Batch input validation (12 tests)
- `TestWebhookConfigFromToml`: Webhook TOML parsing (4 tests)
- `TestSchedulerConfigFromToml`: Scheduler TOML parsing (5 tests)
- `TestArrConfigFromTomlMissingFields`: Missing field handling (6 tests)
- `TestEnvironmentVariableOverrides`: Env var overrides (12 tests)

**Phase 2 - Exception Handling (19 tests):**
- `TestTagApplicationExceptionHandling`: Tag API errors - ConnectError, TimeoutException, ValidationError, HTTPStatusError (7 tests)
- `TestProcessMovieCheckExceptions`: Movie check background task errors (6 tests)
- `TestProcessSeriesCheckExceptions`: Series check background task errors (6 tests)

**Phase 3 - Advanced Features (25 tests):**
- `TestCustomCallableMatcher`: Custom callable matcher for movies/series (8 tests)
- `TestSearchResultBackwardCompat`: `has_4k`, `four_k_releases` properties (3 tests)
- `TestSelectSeasonsDistributedEdgeCases`: DISTRIBUTED strategy edge cases (5 tests)
- `TestImportErrors`: FastAPI/uvicorn import error handling (2 tests)
- `TestSchedulerLifecycle`: Scheduler init, start, and shutdown errors (7 tests)

## Affected Files

| Area | Files | Change Type |
|------|-------|-------------|
| Core | `checker.py` | Granular exception handling |
| Core | `config.py` | Extract parsing helpers |
| Core | `cli.py` | BatchContext + helper extraction |
| Core | `webhook.py` | Granular exception handling |
| Core | `criteria.py` | SPECIAL_EDITION + word boundary matching |
| Tests | `test_scheduler.py` | 85 new tests |
| Tests | `test_webhook.py` | 55 new tests |
| Tests | `test_checker.py` | 24 new tests |
| Tests | `test_cli.py` | 44 new tests |
| Tests | `test_config.py` | 27 new tests |

## New CLI Usage

```bash
# Check a movie for different criteria
filtarr check movie 123 --criteria directors-cut
filtarr check movie 123 --criteria imax
filtarr check movie 123 --criteria special-edition

# Batch check with criteria
filtarr check batch --all-movies --criteria imax
filtarr check batch --all-movies --criteria special-edition
```

## Available Criteria

| Criteria | CLI Name | Applies To | Tag Example |
|----------|----------|------------|-------------|
| 4K/2160p | `4k` | Movies & Series | `4k-available` |
| HDR | `hdr` | Movies & Series | `hdr-available` |
| Dolby Vision | `dolby-vision` | Movies & Series | `dolby-vision-available` |
| Director's Cut | `directors-cut` | Movies only | `directors-cut-available` |
| Extended | `extended` | Movies only | `extended-available` |
| Remaster | `remaster` | Movies only | `remaster-available` |
| IMAX | `imax` | Movies only | `imax-available` |
| Special Edition | `special-edition` | Movies only | `special-edition-available` |

## Commits

| Hash | Message |
|------|---------|
| `db28333` | fix: rewrite import error tests to actually test source code |
| `afbe5d9` | style: fix ruff formatting in test_cli.py |
| `e8a9531` | test: add 115 tests to improve coverage for multi-criteria support |
| `9d79210` | fix: consolidate duplicate import styles in test_webhook.py |
| `da172cd` | fix: address PR review comments |
| `41b11ea` | chore: update uv.lock for v1.1.3 |
| `fc565cb` | test: fix unawaited coroutine warnings and add minor coverage |
| `a9676d5` | test: add webhook error handling and background task coverage |
| `32b5904` | test: add comprehensive scheduler test coverage |
| `26878ce` | refactor(cli): extract batch processing into helper functions |
| `a7c68bf` | refactor(config): extract parsing helpers for improved readability |
| `06a01e4` | fix: add granular exception handling for HTTP, network, and validation errors |
| `9ec999d` | fix: address code review feedback |
| `5a3b60d` | feat: add multi-criteria support beyond 4K |

## Test Plan

- [x] All 466 tests pass
- [x] Ruff lint checks pass
- [x] Mypy type checks pass
- [x] No unawaited coroutine warnings
- [ ] Manual test: `filtarr check movie <id> --criteria directors-cut`
- [ ] Manual test: `filtarr check batch --all-movies --criteria imax --dry-run`
- [ ] Manual test: Verify movie-only criteria rejected for series

---
*Generated with Claude Code*

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -9,18 +9,19 @@
 
 **Running multiple Radarr/Sonarr instances?** You know the problem: your 4K instance fills up with movies that will *never* be available in 4K, your indexers get hammered searching for releases that don't exist, and tools like [Huntarr](https://github.com/plexguide/Huntarr.io) waste API calls hunting for content you'll never find.
 
-**Filtarr solves this.** Instead of blindly syncing your entire library to secondary instances, filtarr checks what's *actually available* from your indexers and tags items accordingly. Only sync what you can actually get.
+**Filtarr solves this.** Instead of blindly syncing your entire library to secondary instances, filtarr checks what's *actually available* from your indexers and tags items accordingly. Filter by 4K, HDR, Director's Cut, IMAX, Special Editions, and more—only sync what you can actually get.
```
Comment (dabigc) [inline README.md]:
```suggestion
**Filtarr solves this.** Instead of blindly syncing your entire library to secondary instances, filtarr checks what's *actually available* from your indexers and tags items accordingly. Filter by 4K, HDR, Director's Cut, IMAX, Special Editions, and more. Only sync what you can actually get.
```

File: README.md [hunk 2]
```diff
@@ -9,18 +9,19 @@
...
-                        └──────────┘              │
-                                                  ▼
+┌─────────────────┐     ┌──────────┐     ┌─────────────────────────┐
+│  Primary Radarr │────▶│  Filtarr │────▶│  Tags Applied           │
+│  (All Movies)   │     │          │     │  • 4k-available         │
+└─────────────────┘     │ Searches │     │  • directors-cut-available│
+                        │ Indexers │     │  • imax-available       │
+                        └──────────┘     └────────────┬────────────┘
+                                                      │
+                                                      ▼
```
Comment (dabigc) [inline README.md]: Add another diagram for Sonarr with an emphasis that it's usually used for 4K with Sonarr.

File: src/filtarr/cli.py [hunk 3]
```diff
@@ -690,13 +810,24 @@ async def run_checks() -> None:
...
+                                    dry_run=dry_run,
                                 )
                     else:
+                        # For series, use 4K if movie-only criteria was specified
+                        # (This handles file-based items with movie-only criteria)
+                        series_criteria = (
+                            SearchCriteria.FOUR_K
+                            if search_criteria in MOVIE_ONLY_CRITERIA
+                            else search_criteria
+                        )
```
[the remark this answers is not in the document]
Comment (dabigc) [reply src/filtarr/cli.py]: ✅ **Addressed in this PR**

Added a warning message when series items in batch files encounter movie-only criteria. The series now gets processed with 4K criteria instead, but the user is explicitly notified:

```
Warning: directors-cut criteria is movie-only. Using 4K for series 'Breaking Bad'
```

This makes the fallback behavior transparent while maintaining the practical workflow for mixed batch files.

Thanks for catching this inconsistency, @Copilot!

File: src/filtarr/config.py [hunk 4]
```diff
@@ -31,13 +31,40 @@ class SonarrConfig:
...
+
+        Returns:
+            Tuple of (available_tag, unavailable_tag)
+        """
+        # Convert underscores to hyphens for tag slugs (e.g., "directors_cut" → "directors-cut")
+        slug = criteria_value.replace("_", "-")
+        return (
+            self.pattern_available.format(criteria=slug),
+            self.pattern_unavailable.format(criteria=slug),
+        )
```
[the remark this answers is not in the document]
Comment (dabigc) [reply src/filtarr/config.py]: ✅ **Addressed in this PR**

Added test coverage for `get_tag_names` in `tests/test_config.py::TestTagConfig`:

**New tests:**
- `test_get_tag_names_with_simple_criteria` - Basic 4k format
- `test_get_tag_names_with_underscore_criteria` - directors_cut → directors-cut conversion
- `test_get_tag_names_with_imax` - Lowercase criteria
- `test_get_tag_names_with_special_edition` - special_edition → special-edition
- `test_get_tag_names_with_custom_pattern` - Custom patterns work correctly
- `test_get_tag_names_with_multiple_underscores` - All underscores converted

Thanks for the suggestion!

File: README.md (hunk 2 above)
Comment (dabigc) [reply README.md]: ✅ Added a separate Sonarr diagram under "### TV Shows (Sonarr)" heading.

The diagram emphasizes that 4K is the primary use case for Sonarr, and includes a note explaining that movie-specific criteria (Director's Cut, IMAX, Special Edition) are not applicable to TV series.

**Comment by dabigc** [comment]
## PR Review Comments Addressed

Commit da172cd addresses the remaining review feedback:

- **docker-compose.yml**: Clarified env_file purpose (container injection vs compose file substitution)
- **publish-pypi.yml**: Explained when to use manual vs automated publishing workflow
- **criteria.py**: Added Args docstring noting title_lower must be pre-lowercased
- **config.py**: Added deprecation timeline (2.0.0) for legacy tag fields
- **tests/test_utils.py**: Extracted duplicate async mock utilities for DRY compliance

File: src/filtarr/criteria.py [hunk 5]
```diff
@@ -119,3 +142,51 @@ def _match_imax(release: Release) -> bool:
...
+        for phrase in (
+            "special edition",
+            "special.edition",
+            "collector's edition",
+            "collectors edition",
+            "collector edition",
+            "anniversary edition",
+            "ultimate edition",
+            "definitive edition",
+        )
```
[the remark this answers is not in the document]
Comment (dabigc) [reply src/filtarr/criteria.py]: The `"special.edition"` variant is **not redundant** and should be kept.

The `_contains_edition_phrase` function uses literal `find()` matching internally:
```python
index = title_lower.find(phrase, start)
```

This means:
- `"special edition"` only matches titles containing the exact substring with a **space**
- `"special.edition"` only matches titles containing the exact substring with a **dot**

The boundary logic (checking for separators before/after the match) validates what appears *around* the phrase, but does not normalize separators *within* the phrase itself.

**Example:**
| Title | `"special edition"` matches? | `"special.edition"` matches? |
|-------|------------------------------|------------------------------|
| `Movie.Special Edition.2160p` | ✅ Yes | ❌ No |
| `Movie.special.edition.2160p` | ❌ No | ✅ Yes |

The test `test_matches_special_edition_with_dots` explicitly validates this behavior with the title `"Movie.2024.special.edition.2160p.BluRay"`.
